Challenge 4 migration to extension

damianmarti opened the pull request:
Challenge 4 migrated to an extension following https://github.com/scaffold-eth/se-2-challenges/issues/234

Install from create-eth repo with:

```
yarn cli -e scaffold-eth/se-2-challenges:ch4-extension-init
```

closes #241

rin-st reviewed (commented):
Thanks Damu! Added small comments

Also found one bug. Steps to reproduce
- add solution contract and also uncomment code on deployment file
- add eth from faucet
- add "1" to "tokenToEth" input. Press "Send"

Result before (text is easy to understand):
<img width="472" alt="Screenshot 2024-12-12 at 19 05 38" src="https://github.com/user-attachments/assets/ec75fc80-513e-4d4b-90d5-d087b516fc23" />

Result in this extension: 
<img width="445" alt="Screenshot 2024-12-12 at 19 07 57" src="https://github.com/user-attachments/assets/593d173d-8126-4db8-9d03-e74d40882548" />

Decoded signature: 
<img width="1039" alt="image" src="https://github.com/user-attachments/assets/c99e30e4-d44f-4c17-aacf-416dfa24a006" />

https://openchain.xyz/signatures?query=0xfb8f41b2

damianmarti commented on extension/packages/hardhat/contracts/Balloons.sol:
```diff
@@ -0,0 +1,10 @@
+    constructor() ERC20("Balloons", "BAL") {
+        _mint(msg.sender, 1000 ether); // mints 1000 balloons!
+    }
+}
```
Added!

damianmarti commented on extension/README.md.args.mjs:
```diff
@@ -0,0 +1,753 @@
+
+📱 Open http://localhost:3000 to see the app.
+
+> 👩‍💻 Rerun \`yarn deploy\` whenever you want to deploy new contracts to the frontend. If you haven't made any contract changes, you can run \`yarn deploy--reset\` for a completely fresh deploy.
```
Fixed!

damianmarti commented on extension/README.md.args.mjs:
```diff
@@ -0,0 +1,753 @@
+  // });
+  \`\`\`
+
+Now, when we \`yarn deploy--reset\` then our contract should be initialized as soon as it deploys, and we should have equal reserves of ETH and tokens.
```
Fixed! And fixed some others I found!

damianmarti commented:
> Thanks Damu! Added small comments
> 
> Also found one bug. Steps to reproduce
> 
> * add solution contract and also uncomment code on deployment file
> * add eth from faucet
> * add "1" to "tokenToEth" input. Press "Send"
> 
> Result before (text is easy to understand): <img alt="Screenshot 2024-12-12 at 19 05 38" width="472" src="https://private-user-images.githubusercontent.com/25638585/395280905-ec75fc80-513e-4d4b-90d5-d087b516fc23.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQwOTk3NjksIm5iZiI6MTczNDA5OTQ2OSwicGF0aCI6Ii8yNTYzODU4NS8zOTUyODA5MDUtZWM3NWZjODAtNTEzZS00ZDRiLTkwZDUtZDA4N2I1MTZmYzIzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjEzVDE0MTc0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFlZWRkMjliZWRmYmI4OGRiMzk4NTE3YzUyNzIzMzQ1OGM4Zjk0ODI0MjY1OGVjNzBmYjYwNjBkYzU4NjczNWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.H9crCwvOXlyB3cipuYzMjt9-lr8BofyC6a4NSz6ub-Y">
> 
> Result in this extension: <img alt="Screenshot 2024-12-12 at 19 07 57" width="445" src="https://private-user-images.githubusercontent.com/25638585/395281090-593d173d-8126-4db8-9d03-e74d40882548.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQwOTk3NjksIm5iZiI6MTczNDA5OTQ2OSwicGF0aCI6Ii8yNTYzODU4NS8zOTUyODEwOTAtNTkzZDE3M2QtODEyNi00ZGI4LTlkMDMtZTc0ZDQwODgyNTQ4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjEzVDE0MTc0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThkNGY5ODE5MjUxYzRhYmIxYWUxMjZmZmYzNDk2ZjRkYjlhM2I0MmE0NTljYmY2YTZmM2EwZDU3ODA2ZDVkMzMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.AGz4-0-C5AlVZC-4cg50OmHSO7xxSUsr2Cg_iyIXIQA">
> 
> Decoded signature: <img alt="image" width="1039" src="https://private-user-images.githubusercontent.com/25638585/395281663-c99e30e4-d44f-4c17-aacf-416dfa24a006.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQwOTk3NjksIm5iZiI6MTczNDA5OTQ2OSwicGF0aCI6Ii8yNTYzODU4NS8zOTUyODE2NjMtYzk5ZTMwZTQtZDQ0Zi00YzE3LWFhY2YtNDE2ZGZhMjRhMDA2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjEzVDE0MTc0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTlmZjQ1ZTZlZjNmYWMyZDE3NDQ4NDk3ZDYzN2JlNDg5NjMzZGFiOWUyMDliZDgwYjA1MGNiNjM4ZGE4ZGEwODAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.E0iT_dGjZBEHSnn0OWbPKSTHMSqEKOsrT4rjrilXUQg">
> 
> https://openchain.xyz/signatures?query=0xfb8f41b2

Thanks @rin-st for the review and great catch!

I fixed it by adding a package.json to hardhat with @openzeppelin/contracts.

The new @openzeppelin/contracts v5 uses custom errors and our getParsedError is not able to show a nice message in this case. We should add an issue to fix it.

rin-st commented:
Lgtm! Created an issue to fix that error, see above.

Thank you @damianmarti !

Pabl0cks commented on extension/README.md.args.mjs:
```diff
@@ -0,0 +1,753 @@
+
+Navigate to the \`Debug Contracts\` tab, you should see two smart contracts displayed called \`DEX\` and \`Balloons\`.
+
+\`packages / hardhat / contracts / Balloons.sol\` is just an example ERC20 contract that mints 1000 $BAL to whatever address deploys it.
```
Need to delete the spaces in this route

Pabl0cks commented on extension/README.md.args.mjs:
```diff
@@ -0,0 +1,753 @@
+
+\`packages / hardhat / contracts / Balloons.sol\` is just an example ERC20 contract that mints 1000 $BAL to whatever address deploys it.
+
+\`packages / hardhat / contracts / DEX.sol\` is what we will build in this challenge and you can see it starts instantiating a token (ERC20 interface) that we set in the constructor (on deploy).
```
Same

Pabl0cks commented on extension/README.md.args.mjs:
```diff
@@ -0,0 +1,753 @@
+
+### ⚔️ Side Quests
+
+- [ ] In \`packages\nextjs\app\events\page.tsx\` implement an event and emit for the \`approve()\` function to make it clear when it has been executed.
```
\`packages\nextjs\app\events\page.tsx\`  this is getting rendered badly in my instance readme:

```
In `packages
extjsappeventspage.tsx`
```

Pabl0cks replied:
I think there was a backslash escaping problem here, was still seeing it wrong. 

As a quick fix just changed the backslash for regular slash in https://github.com/scaffold-eth/se-2-challenges/pull/245/commits/a9a4a12be7e17e76accf01c9388e3b649799e77b since we're using regular slashes in other folders.

Pabl0cks commented on extension/README.md.args.mjs:
```diff
@@ -0,0 +1,753 @@
+
+> Follow the steps to deploy to Vercel. Once you log in (email, github, etc), the default options should work. It'll give you a public URL.
+
+> If you want to redeploy to the same production URL you can run \`yarn vercel--prod\`. If you omit the \`--prod\` flag it will deploy it to a preview/test URL.
```
Missing a space in `yarn vercel--prod`

Pabl0cks reviewed (commented):
Nice job Damu!! 🙌 The challenge is working great to me.

Just added some comments with nitpick format errors in the Readme from my instance.

Also just to comment it, I received this message when I tried to `yarn vercel`. Never saw it before, I needed to do `yarn vercel --yes`:

```
$ yarn vercel
Vercel CLI 39.1.3
> NOTE: The Vercel CLI now collects telemetry regarding usage of the CLI.
> This information is used to shape the CLI roadmap and prioritize features.
> You can learn more, including how to opt-out if you'd not like to participate in this program, by visiting the following URL:
> https://vercel.com/docs/cli/about-telemetry
Error: Command `vercel deploy` requires confirmation. Use option "--yes" to confirm.
```

rin-st commented:
> Also just to comment it, I received this message when I tried to yarn vercel. Never saw it before, I needed to do yarn vercel --yes:

Could you check if it asking it for new deployments? Probably you need to create new instance

Also, try it with env variables like in this [PR](https://github.com/scaffold-eth/scaffold-eth-2/pull/1012)

Pabl0cks commented:
> Could you check if it asking it for new deployments? Probably you need to create new instance, or is it asked just one time?

>
> Also, try it with env variables like in this https://github.com/scaffold-eth/scaffold-eth-2/pull/1012

Just created a fresh instance of another extension and seems like today I'm getting another error, without the telemetry promt:

![image](https://github.com/user-attachments/assets/3bf89981-2961-4377-9915-a4e952f42988)

Tried with [PR](https://github.com/scaffold-eth/scaffold-eth-2/pull/1012) but still get the message, I needed to add the `--yes` in order to make it work

rin-st commented:
from the [docs](https://vercel.com/docs/cli/deploy#yes) :

`The --yes option can be used to skip questions you are asked when setting up a new Vercel project. The questions will be answered with the provided defaults, inferred from vercel.json and the folder name.`

So as I understand it should be a question before asking for `--yes`. Looks like a vercel cli bug.

---

Could you please try one more thing for the last project:
```
yarn vercel telemetry disable
```
and then 
```
yarn vercel
```
Is it still asking for `--yes`?

damianmarti commented:
> Nice job Damu!! 🙌 The challenge is working great to me.
> 
> Just added some comments with nitpick format errors in the Readme from my instance.

Thanks @Pabl0cks for the review!!

I used a clean README and pushed the changes. There was an issue with the previous one due to auto-formatting from my IDE. Please, let me know if the formatting is ok now.

Pabl0cks commented:
> I'm using Vercel CLI 37.4.2 and it's working fine. Maybe some wrong project configuration. Can you please try to do it from a fresh project?

Realized it was only happening to my Git Bash (on desktop). The rest of terminals were working well, even Git Bash on laptop.
Downloaded and installed last version of Git Bash and is working nicely on desktop without `--yes` prompt🙏

Pabl0cks reviewed (approved):
Looking good to merge! TY @damianmarti
